File: edge-orch/state-aggregator/tests/test_common_runtime_demo.js
const test=require('node:test'),assert=require('node:assert/strict');
const D=require('../app/static/nexus/common-runtime-demo.js');
const data={enabled:true,observedAt:99,items:[{uid:'u',name:'svc',label:'Synthetic',available:true,maxRequests:512,concurrency:6}],runs:[]};
test('actions are tied to opted-in identity and disabled on stale/pending/running state',()=>{
 assert.equal(D.actions(data,'unknown',true,false),'');
 assert.match(D.actions(data,'u',true,false),/시험 요청 1건/);
 assert.doesNotMatch(D.actions(data,'u',true,false),/disabled|token|토큰 입력/);
 assert.match(D.actions(data,'u',false,false),/disabled/);
 assert.match(D.actions(data,'u',true,true),/disabled/);
 assert.match(D.actions({...data,runs:[{uid:'u',phase:'Running'}]},'u',true,false),/disabled/);
});
test('unknown acceptance retains same-ID action; failed round trip does not claim success',()=>{
 const run={id:'run',name:'<svc>',label:'synthetic',mode:'round-trip',phase:'Incomplete',createdAt:99,
  sent:2,succeeded:2,failed:0,unknown:0,returned:false,retiring:0,routeHistory:[],lastResult:'<script>bad</script>'};
 const html=D.panel({...data,runs:[run]},null,[{name:'svc',uid:'u'}],100);
 assert.match(html,/같은 실행 ID로 확인·재접수/);
 assert.match(html,/검증 미완료/);assert.match(html,/역할 복귀 미확인/);
 assert.match(html,/&lt;script&gt;/);assert.doesNotMatch(html,/<script>/);
});
test('plain HTTP UUID fallback needs no secure-context randomUUID',()=>{
 let count=0;const crypto={getRandomValues:bytes=>{bytes.fill(++count);return bytes;}};
 const first=D.newId(crypto),second=D.newId(crypto);
 assert.match(first,/^[0-9a-f]{32}$/);assert.notEqual(first,second);
});

test('approval-mode AI load is distinct from automatic round-trip',()=>{
 const html=D.actions(data,'u',true,false,true);
 assert.match(html,/data-demo-mode="load"/);assert.match(html,/AI 추론 부하 주기/);
 assert.doesNotMatch(html,/왕복 시험/);
 const run={id:'load',name:'AI',label:'fixed AI',mode:'load',phase:'Completed',createdAt:99,
 sent:20,succeeded:20,failed:0,unknown:0,returned:false,retiring:0,routeHistory:[]};
 assert.match(D.panel({...data,runs:[run]},null,[],100),/부하 시험 완료/);
 assert.doesNotMatch(D.panel({...data,runs:[run]},null,[],100),/검증 통과/);
});

test('running load has immediate local controls and honest indeterminate progress',()=>{
 const run={uid:'u',id:'r',name:'AI',mode:'load',phase:'Running',stage:'pressure',createdAt:90,sent:20,succeeded:18,failed:1,unknown:0};
 const buttons=D.actions({...data,runs:[run]},'u',true,false,true);assert.match(buttons,/실행 중…/);assert.match(buttons,/aria-busy="true"/);
 const html=D.runStatus({current:true,run},100);assert.match(html,/집중 부하 실행 중/);assert.match(html,/10초/);assert.match(html,/새 시험 요청 중단/);assert.match(html,/role="progressbar"/);assert.doesNotMatch(html,/aria-valuenow/);
 const stale=D.runStatus({current:false,run},200);assert.match(stale,/실행 상태 확인 불가/);assert.doesNotMatch(stale,/runtime-spinner|data-demo-stop|data-runtime-elapsed/);
 const stopping=D.runStatus({current:true,run:{...run,phase:'Stopping'}},100);assert.match(stopping,/요청 마무리/);assert.match(stopping,/disabled/);
 const sending=D.runStatus({current:true,sending:{mode:'load'}},100);assert.match(sending,/시작 요청 접수 중/);
 const failed=D.runStatus({current:true,run:{...run,phase:'Incomplete',finishedAt:98}},100);assert.match(failed,/검증 미완료/);assert.doesNotMatch(failed,/runtime-spinner|부하 시험 완료/);
});

test('load stop is always visible beside start, gated by exact running identity',()=>{
 let html=D.actions(data,'u',true,false,true);assert.match(html,/부하 제거/);assert.doesNotMatch(html,/data-demo-stop=/);
 const run={uid:'u',id:'exact-run',name:'svc',mode:'load',phase:'Running'};
 html=D.actions({...data,runs:[run]},'u',true,false,true);assert.match(html,/data-demo-stop="exact-run"/);assert.match(html,/id="demo-stop-control-u"/);
 const stop=html.match(/<button class="button runtime-stop"[^>]*>/)[0];assert.doesNotMatch(stop,/disabled/);
 html=D.actions({...data,runs:[{...run,phase:'Stopping'}]},'u',true,false,true);assert.match(html,/부하 제거 중/);assert.match(html.match(/<button class="button runtime-stop"[^>]*>/)[0],/disabled/);
 html=D.runStatus({current:true,inlineControls:true,run:{...run,createdAt:90,sent:2,succeeded:1,failed:0,unknown:0}},100);assert.doesNotMatch(html,/data-demo-stop=/);assert.match(html,/부하 제거:/);
});


test('service start and stop are explicit lifecycle controls, separate from node load',()=>{
 const item={...data.items[0],serviceControl:{phase:'Running',canStart:false,canStop:true}};
 let html=D.actions({...data,items:[item]},'u',true,false,true);
 assert.match(html,/서비스 실행/);assert.match(html,/서비스 중지/);assert.doesNotMatch(html,/data-demo-mode="load"/);
 assert.match(html.match(/<button id="service-start-u"[^>]*>/)[0],/disabled/);
 assert.doesNotMatch(html.match(/<button id="service-stop-u"[^>]*>/)[0],/disabled/);
 item.serviceControl={phase:'Stopped',canStart:true,canStop:false};
 html=D.serviceActions({...data,items:[item]},'u',true);
 assert.doesNotMatch(html.match(/<button id="service-start-u"[^>]*>/)[0],/disabled/);
 assert.match(html,/서비스 중지됨/);
 html=D.serviceActions({...data,items:[item]},'u',false);
 assert.match(html,/서비스 상태 확인 불가/);assert.match(html.match(/<button id="service-start-u"[^>]*>/)[0],/disabled/);
});

test('each node has its own paired controls and only the current route accepts load',()=>{
 const value={...data,items:[{...data.items[0],nodes:[{node:'nano',label:'Nano',available:true,currentRoute:true},{node:'orin',label:'Orin',available:false,currentRoute:false}]}]};
 let nano=D.nodeActions(value,'u','nano','Nano',true),orin=D.nodeActions(value,'u','orin','Orin',true);
 assert.match(nano,/Nano에 부하 주기/);assert.match(nano,/Nano 부하 제거/);assert.match(nano,/data-demo-target-node="nano"/);
 assert.doesNotMatch(nano.match(/<button id="node-load-u-nano"[^>]*>/)[0],/disabled/);
 assert.match(orin.match(/<button id="node-load-u-orin"[^>]*>/)[0],/disabled/);
 assert.match(orin,/서비스가 이 노드에서 실행되면 사용/);
 value.runs=[{uid:'u',id:'nano-run',name:'svc',targetNode:'nano',phase:'Running',sent:3,succeeded:1,cancelled:0,failed:0}];
 nano=D.nodeActions(value,'u','nano','Nano',true);orin=D.nodeActions(value,'u','orin','Orin',true);
 assert.match(nano,/data-demo-stop="nano-run"/);assert.doesNotMatch(orin,/data-demo-stop=/);
 assert.match(nano,/Nano 부하 중/);assert.match(nano,/role="progressbar"/);
 value.runs[0]={...value.runs[0],phase:'Stopped',cancelled:2,reason:'node_route_changed'};
 nano=D.nodeActions(value,'u','nano','Nano',true);assert.match(nano,/실행 노드 이동/);assert.match(nano,/취소 2/);
 const stale=D.nodeActions(value,'u','nano','Nano',false);assert.doesNotMatch(stale,/runtime-spinner|role="progressbar"/);assert.match(stale,/상태 확인 불가/);
});

File: edge-orch/state-aggregator/tests/test_common_runtime_demo.py
import asyncio
from types import SimpleNamespace

import httpx
from fastapi import FastAPI

from app.common_runtime_demo import create_common_demo_router


def receipt():
    return {"id": "run-one1", "uid": "service-uid", "name": "service", "label": "Fixed demo",
        "mode": "single", "phase": "Running", "stage": "single", "createdAt": 100,
        "sent": 0, "succeeded": 0, "failed": 0, "unknown": 0, "lastResult": None,
        "routeHistory": [], "recentRequests": [], "startNode": "any-edge", "startRole": "edge",
        "returned": False, "retiring": 0, "stopRequested": False}


def test_token_free_same_origin_proxy_restricts_mutations_to_declared_run_identity():
    async def run():
        calls = []
        def handler(request):
            calls.append(request)
            return httpx.Response(202, json=receipt())
        app = FastAPI()
        app.include_router(create_common_demo_router(SimpleNamespace(common_runtime_demo_enabled=True,
            common_runtime_url="http://operator"), transport=httpx.MockTransport(handler)))
        async with httpx.AsyncClient(transport=httpx.ASGITransport(app=app), base_url="http://nexus") as client:
            path = "/api/runtime-demos/service/runs/run-one1"
            body = {"serviceUid": "service-uid", "mode": "single"}
            good = {"Origin": "http://nexus", "X-Runtime-Demo": "1"}
            for headers in ({}, {"Origin": "http://evil", "X-Runtime-Demo": "1"}, {"Origin": "http://nexus"}):
                assert (await client.post(path, json=body, headers=headers)).status_code == 403
            assert not calls
            assert (await client.post(path, json={**body, "payload": {"arbitrary": True}}, headers=good)).status_code == 422
            response = await client.post(path, json=body, headers=good)
            assert response.status_code == 202 and response.json()["phase"] == "Running"
            assert len(calls) == 1 and calls[0].url == "http://operator/demos/service/runs/run-one1"
            assert "execution-token" not in str(calls[0].headers).lower()
            assert (await client.post(path+"/stop", json={"serviceUid": "service-uid"}, headers=good)).status_code == 200
            assert calls[-1].url.path.endswith("/stop")
    asyncio.run(run())


def test_disabled_demo_and_source_failures_are_explicit_without_following_redirects():
    async def run():
        for mode in ("disabled", "offline", "redirect", "malformed"):
            calls = []
            def handler(request):
                calls.append(request)
                if mode == "offline":
                    raise httpx.ConnectError("private-url")
                if mode == "redirect":
                    return httpx.Response(302, headers={"Location": "http://other"}, json={})
                return httpx.Response(500, json=[])
            app = FastAPI()
            app.include_router(create_common_demo_router(SimpleNamespace(common_runtime_demo_enabled=mode!="disabled",
                common_runtime_url="http://operator"), transport=httpx.MockTransport(handler), clock=lambda:100))
            async with httpx.AsyncClient(transport=httpx.ASGITransport(app=app), base_url="http://nexus") as client:
                response = await client.get("/state/runtime-demos")
                assert response.headers["Cache-Control"] == "no-store"
                data = response.json()
                if mode == "disabled":
                    assert not data["enabled"] and not calls
                else:
                    assert data["observation_error"] == "demo_source_unavailable" and len(calls) == 1
                    assert "private" not in response.text
    asyncio.run(run())


def test_augmentation_approval_preserves_identity_and_requires_same_origin():
    async def run():
        calls = []
        def handler(request):
            calls.append(request)
            return httpx.Response(202, json={"id":"a"*32,"createdAt":100,"expiresAt":160,
                "sourceNode":"edge","node":"server","role":"server","variant":"model",
                "reason":"sustained_pressure","status":"Approved"})
        app=FastAPI()
        app.include_router(create_common_demo_router(SimpleNamespace(common_runtime_demo_enabled=True,
            common_runtime_url="http://operator"),transport=httpx.MockTransport(handler)))
        async with httpx.AsyncClient(transport=httpx.ASGITransport(app=app),base_url="http://nexus") as client:
            path="/api/runtime-demos/llama/augmentation/approve"
            body={"serviceUid":"service-uid","recommendationId":"a"*32}
            assert (await client.post(path,json=body)).status_code==403
            assert not calls
            response=await client.post(path,json=body,headers={"Origin":"http://nexus","X-Runtime-Demo":"1"})
            assert response.status_code==202 and response.json()["status"]=="Approved"
            assert calls[0].url.path=="/services/llama/augmentation/approve"
    asyncio.run(run())
